src/Interpreters/Session.cpp

```diff
@@ -214,55 +223,48 @@ class NamedSessionsStorage
         std::unique_lock lock{mutex};
         while (!quit)
         {
-            auto interval = closeSessions(lock);
-            if (cond.wait_for(lock, interval, [this]() -> bool { return quit; }))
+            closeSessions(lock);
+            if (cond.wait_for(lock, close_interval, [this]() -> bool { return quit; }))
                 break;
         }
     }
 
-    /// Close sessions, that has been expired. Returns how long to wait for next session to be expired, if no new sessions will be added.
-    std::chrono::steady_clock::duration closeSessions(std::unique_lock<std::mutex> & lock)
+    void closeSessions(std::unique_lock<std::mutex> & lock)
     {
         const auto now = std::chrono::steady_clock::now();
 
-        /// The time to close the next session did not come
-        if (now < close_cycle_time)
-            return close_cycle_time - now;  /// Will sleep until it comes.
-
-        const auto current_cycle = close_cycle;
-
-        ++close_cycle;
-        close_cycle_time = now + close_interval;
+        while (!close_time_buckets.empty())
+        {
+            const auto & [time_bucket, session_keys] = *close_time_buckets.begin();
+            if (time_bucket > now)
+                break;
 
-        if (close_times.empty())
-            return close_interval;
+            for (const auto & key : session_keys)
+            {
+                const auto & session_it = sessions.find(key);
 
-        auto & sessions_to_close = close_times.front();
+                if (session_it == sessions.end())
+                    continue;
 
-        for (const auto & key : sessions_to_close)
-        {
-            const auto session = sessions.find(key);
+                const auto & session = session_it->second;
 
-            if (session != sessions.end() && session->second->close_cycle <= current_cycle)
```

Comment:
here was the issue, more precisely in `session->second->close_cycle <= current_cycle`
when we delay closing because of count higher than 1 we set timeout to 0 and schedule it again.
The problem is that we will use **new** `close_cycle` value to set `session.close_cycle`
but that same session will be put into `close_times` relative to the **old** `close_cycle`